In [1]:
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers.core import Dense, Activation
from keras.optimizers import SGD
from keras.utils import np_utils
np.random.seed(1671)

Using TensorFlow backend.
E:\Miniconda3\envs\deeplearning\lib\site-packages\tensorflow\python\framework\dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
E:\Miniconda3\envs\deeplearning\lib\site-packages\tensorflow\python\framework\dtypes.py:517: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
E:\Miniconda3\envs\deeplearning\lib\site-packages\tensorflow\python\framework\dtypes.py:518: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
E:\Miniconda3\envs\deeplearning\lib\site-packages\tensorflo

In [5]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

11493376/11490434 [==============================] - 10s 1us/step


In [6]:
X_train.shape

(60000, 28, 28)

In [7]:
NP_EPOCH = 200
BATCH_SIZE = 128
VERBOSE = 1
NB_CLASSES = 10 #the no. of neurons in the final layer (multiclass classsification)
OPTIMIZER = SGD() 
N_HIDDEN = 128
VALIDATION_SPLIT = 0.2
RESHAPED = 784

In [9]:
#reshaping the 28x28 to 784
X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype("float32")
X_test = X_test.astype("float32")

In [10]:
#normalizing
X_train /= 255
X_test /= 255

In [12]:
print("training examples: ", str(X_train.shape[0]))
print("testing examples: ", str(X_test.shape[0]))

training examples:  60000
testing examples:  10000


In [17]:
#convert class vectors to binary class matrices
y_train = np_utils.to_categorical(y_train, NB_CLASSES)
y_test = np_utils.to_categorical(y_test, NB_CLASSES)

In [18]:
#input layer will have 28x28 = 784 neurons
# there will be 10 neurons in the output layer representing the 10 numbers
#softmax function is used for multiclass classification
#softmax squashes the values in range(0,1)
model = Sequential()
model.add(Dense(NB_CLASSES, input_shape=(RESHAPED,)))
model.add(Activation("softmax"))
model.summary()




_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 10)                7850      
_________________________________________________________________
activation_1 (Activation)    (None, 10)                0         
Total params: 7,850
Trainable params: 7,850
Non-trainable params: 0
_________________________________________________________________


In [19]:
#categorical_crossentropy is  the multiclass logarithmic loss
model.compile(loss="categorical_crossentropy", optimizer=OPTIMIZER, metrics=["accuracy"])


Instructions for updating:
keep_dims is deprecated, use keepdims instead



In [20]:
#we reserve a part of the training data for measuring performance
# the validation while training
history = model.fit(X_train, y_train, batch_size=BATCH_SIZE,
                    epochs= NP_EPOCH, verbose=VERBOSE,
                    validation_split=VALIDATION_SPLIT)

Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where

Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
Train on 48000 samples, validate on 12000 samples
Epoch 1/200
48000/48000 [==============================] - 2s 46us/step - loss: 1.3633 - acc: 0.6796 - val_loss: 0.8904 - val_acc: 0.8246
Epoch 2/200
48000/48000 [==============================] - 1s 24us/step - loss: 0.7913 - acc: 0.8272 - val_loss: 0.6572 - val_acc: 0.8546
Epoch 3/200
48000/48000 [==============================] - 1s 23us/step - loss: 0.6436 - acc: 0.8497 - val_loss: 0.5625 - val_acc: 0.8681
Epoch 4/200
48000/48000 [==============================] - 1s 24us/step - loss: 0.5717 - acc: 0.8602 - val_loss: 0.5098 - val_acc: 0.8765
Epoch 5/200
48000/48000 [==============================] - 1s 26us/step - loss: 0.5276 - acc: 0.8678 - val_loss: 0.4758 - val_acc: 0.8826
Epoch 6/200
48000/48000 [======================

48000/48000 [==============================] - 1s 25us/step - loss: 0.3187 - acc: 0.9114 - val_loss: 0.3057 - val_acc: 0.9153
Epoch 54/200
48000/48000 [==============================] - 1s 24us/step - loss: 0.3180 - acc: 0.9117 - val_loss: 0.3050 - val_acc: 0.9148
Epoch 55/200
48000/48000 [==============================] - 1s 27us/step - loss: 0.3171 - acc: 0.9121 - val_loss: 0.3044 - val_acc: 0.9149
Epoch 56/200
48000/48000 [==============================] - 1s 25us/step - loss: 0.3164 - acc: 0.9121 - val_loss: 0.3037 - val_acc: 0.9156
Epoch 57/200
48000/48000 [==============================] - 1s 29us/step - loss: 0.3157 - acc: 0.9128 - val_loss: 0.3034 - val_acc: 0.9152
Epoch 58/200
48000/48000 [==============================] - 1s 23us/step - loss: 0.3149 - acc: 0.9121 - val_loss: 0.3029 - val_acc: 0.9148
Epoch 59/200
48000/48000 [==============================] - 1s 23us/step - loss: 0.3143 - acc: 0.9128 - val_loss: 0.3022 - val_acc: 0.9151
Epoch 60/200
48000/48000 [==============

48000/48000 [==============================] - 1s 19us/step - loss: 0.2921 - acc: 0.9184 - val_loss: 0.2858 - val_acc: 0.9210
Epoch 112/200
48000/48000 [==============================] - 1s 20us/step - loss: 0.2918 - acc: 0.9187 - val_loss: 0.2857 - val_acc: 0.9207
Epoch 113/200
48000/48000 [==============================] - 1s 21us/step - loss: 0.2915 - acc: 0.9189 - val_loss: 0.2854 - val_acc: 0.9210
Epoch 114/200
48000/48000 [==============================] - 1s 20us/step - loss: 0.2913 - acc: 0.9188 - val_loss: 0.2853 - val_acc: 0.9211
Epoch 115/200
48000/48000 [==============================] - 1s 20us/step - loss: 0.2910 - acc: 0.9189 - val_loss: 0.2852 - val_acc: 0.9205
Epoch 116/200
48000/48000 [==============================] - 1s 19us/step - loss: 0.2908 - acc: 0.9189 - val_loss: 0.2849 - val_acc: 0.9213
Epoch 117/200
48000/48000 [==============================] - 1s 19us/step - loss: 0.2905 - acc: 0.9193 - val_loss: 0.2847 - val_acc: 0.9213
Epoch 118/200
48000/48000 [=======

48000/48000 [==============================] - 1s 20us/step - loss: 0.2803 - acc: 0.9216 - val_loss: 0.2782 - val_acc: 0.9227
Epoch 170/200
48000/48000 [==============================] - ETA: 0s - loss: 0.2795 - acc: 0.921 - 1s 19us/step - loss: 0.2801 - acc: 0.9216 - val_loss: 0.2781 - val_acc: 0.9227
Epoch 171/200
48000/48000 [==============================] - 1s 20us/step - loss: 0.2800 - acc: 0.9220 - val_loss: 0.2780 - val_acc: 0.9226
Epoch 172/200
48000/48000 [==============================] - 1s 19us/step - loss: 0.2798 - acc: 0.9218 - val_loss: 0.2778 - val_acc: 0.9231
Epoch 173/200
48000/48000 [==============================] - 1s 19us/step - loss: 0.2797 - acc: 0.9217 - val_loss: 0.2778 - val_acc: 0.9229
Epoch 174/200
48000/48000 [==============================] - 1s 20us/step - loss: 0.2796 - acc: 0.9217 - val_loss: 0.2777 - val_acc: 0.9227
Epoch 175/200
48000/48000 [==============================] - 1s 19us/step - loss: 0.2794 - acc: 0.9218 - val_loss: 0.2776 - val_acc: 0.9

In [21]:
score = model.evaluate(X_test, y_test, verbose=VERBOSE)
print("test score: ", score[0])
print("test accuracy: ", score[1])

10000/10000 [==============================] - 0s 45us/step
test score:  0.2773858502626419
test accuracy:  0.9227


Hence, we achieved an accuracy of 92.27% on the test data.

### Improving the model with hidden layers

In [22]:
NP_EPOCH = 20
model2 = Sequential()

In [23]:
model2.add(Dense(N_HIDDEN, input_shape = (RESHAPED,)))
model2.add(Activation('relu'))

In [24]:
#the 1st hidden layer
model2.add(Dense(N_HIDDEN))
model2.add(Activation('relu'))

In [26]:
model2.add(Dense(NB_CLASSES))
model2.add(Activation('softmax'))

In [30]:
model2.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_2 (Dense)              (None, 128)               100480    
_________________________________________________________________
activation_2 (Activation)    (None, 128)               0         
_________________________________________________________________
dense_3 (Dense)              (None, 128)               16512     
_________________________________________________________________
activation_3 (Activation)    (None, 128)               0         
_________________________________________________________________
dense_4 (Dense)              (None, 10)                1290      
_________________________________________________________________
dense_5 (Dense)              (None, 10)                110       
_________________________________________________________________
activation_5 (Activation)    (None, 10)                0         
Total para

In [31]:
model2.compile(loss = 'categorical_crossentropy', optimizer = OPTIMIZER, metrics = ['accuracy'])

In [32]:
model2.fit(X_train, y_train, batch_size=BATCH_SIZE, epochs=NP_EPOCH, verbose=VERBOSE, validation_split=VALIDATION_SPLIT)

Train on 48000 samples, validate on 12000 samples
Epoch 1/20
48000/48000 [==============================] - 2s 46us/step - loss: 1.4975 - acc: 0.5817 - val_loss: 0.8414 - val_acc: 0.7628
Epoch 2/20
48000/48000 [==============================] - 2s 36us/step - loss: 0.6772 - acc: 0.8053 - val_loss: 0.5155 - val_acc: 0.8525
Epoch 3/20
48000/48000 [==============================] - 2s 39us/step - loss: 0.4747 - acc: 0.8661 - val_loss: 0.3968 - val_acc: 0.8856
Epoch 4/20
48000/48000 [==============================] - 2s 39us/step - loss: 0.3860 - acc: 0.8902 - val_loss: 0.3383 - val_acc: 0.9022
Epoch 5/20
48000/48000 [==============================] - 2s 37us/step - loss: 0.3352 - acc: 0.9039 - val_loss: 0.3010 - val_acc: 0.9113
Epoch 6/20
48000/48000 [==============================] - 2s 41us/step - loss: 0.3003 - acc: 0.9137 - val_loss: 0.2748 - val_acc: 0.9197
Epoch 7/20
48000/48000 [==============================] - 2s 43us/step - loss: 0.2742 - acc: 0.9209 - val_loss: 0.2529 - val_acc

In [34]:
score = model2.evaluate(X_test, y_test, verbose=VERBOSE)
print("test score: ", score[0])
print("test accuracy: ", score[1])

10000/10000 [==============================] - 0s 45us/step
test score:  0.14248205675780773
test accuracy:  0.9578


Improved accuracy: new accuracy is 95.78% 

### Experimenting with dropout 

In [35]:
DROPOUT = 0.3

In [37]:
model3 = Sequential()

In [40]:
from keras.layers import Dropout

In [42]:
model3.add(Dropout(DROPOUT, input_shape = (RESHAPED,)))
model3.add(Dense(N_HIDDEN, activation='relu'))

In [44]:
model3.add(Dense(NB_CLASSES, activation='softmax'))

In [45]:
model3.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_7 (Dense)              (None, 128)               100480    
_________________________________________________________________
activation_7 (Activation)    (None, 128)               0         
_________________________________________________________________
dropout_1 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_8 (Dense)              (None, 128)               16512     
_________________________________________________________________
dropout_2 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_9 (Dense)              (None, 128)               16512     
_________________________________________________________________
dense_10 (Dense)             (None, 10)                1290      
Total para

In [46]:
model3.compile(loss = 'categorical_crossentropy', optimizer = OPTIMIZER, metrics = ['accuracy'])

In [47]:
model3.fit(X_train, y_train, batch_size=BATCH_SIZE, epochs=NP_EPOCH, verbose=VERBOSE, validation_split=VALIDATION_SPLIT)

Train on 48000 samples, validate on 12000 samples
Epoch 1/20
48000/48000 [==============================] - 3s 66us/step - loss: 1.8897 - acc: 0.4003 - val_loss: 1.0759 - val_acc: 0.7549
Epoch 2/20
48000/48000 [==============================] - 3s 54us/step - loss: 1.0297 - acc: 0.6713 - val_loss: 0.5841 - val_acc: 0.8461
Epoch 3/20
48000/48000 [==============================] - 3s 56us/step - loss: 0.7495 - acc: 0.7680 - val_loss: 0.4472 - val_acc: 0.8794
Epoch 4/20
48000/48000 [==============================] - 3s 59us/step - loss: 0.6244 - acc: 0.8104 - val_loss: 0.3794 - val_acc: 0.8937
Epoch 5/20
48000/48000 [==============================] - 3s 62us/step - loss: 0.5473 - acc: 0.8350 - val_loss: 0.3376 - val_acc: 0.9040
Epoch 6/20
48000/48000 [==============================] - 4s 73us/step - loss: 0.4938 - acc: 0.8527 - val_loss: 0.3105 - val_acc: 0.9087
Epoch 7/20
48000/48000 [==============================] - 4s 74us/step - loss: 0.4583 - acc: 0.8642 - val_loss: 0.2863 - val_acc

In [48]:
score = model3.evaluate(X_test, y_test, verbose=VERBOSE)
print("test score: ", score[0])
print("test accuracy: ", score[1])

10000/10000 [==============================] - 1s 59us/step
test score:  0.17060165482982992
test accuracy:  0.9479


Accuracy slightly lower when dropout has been applied. 

### Let us change the optimizer to RMSProp and try again.

In [49]:
from keras.optimizers import RMSprop

In [51]:
OPTIMIZER = RMSprop()
model3.compile(loss = 'categorical_crossentropy', optimizer = OPTIMIZER, metrics = ['accuracy'])

In [52]:
model3.fit(X_train, y_train, batch_size=BATCH_SIZE, epochs=NP_EPOCH, verbose=VERBOSE, validation_split=VALIDATION_SPLIT)

Train on 48000 samples, validate on 12000 samples
Epoch 1/20
48000/48000 [==============================] - 3s 72us/step - loss: 0.2469 - acc: 0.9257 - val_loss: 0.1385 - val_acc: 0.9604
Epoch 2/20
48000/48000 [==============================] - 3s 63us/step - loss: 0.1845 - acc: 0.9454 - val_loss: 0.1144 - val_acc: 0.9671
Epoch 3/20
48000/48000 [==============================] - 3s 59us/step - loss: 0.1541 - acc: 0.9548 - val_loss: 0.1092 - val_acc: 0.9693
Epoch 4/20
48000/48000 [==============================] - 3s 65us/step - loss: 0.1385 - acc: 0.9587 - val_loss: 0.0983 - val_acc: 0.9720
Epoch 5/20
48000/48000 [==============================] - 3s 54us/step - loss: 0.1248 - acc: 0.9628 - val_loss: 0.0970 - val_acc: 0.9737
Epoch 6/20
48000/48000 [==============================] - 3s 63us/step - loss: 0.1126 - acc: 0.9667 - val_loss: 0.0970 - val_acc: 0.9731
Epoch 7/20
48000/48000 [==============================] - 3s 67us/step - loss: 0.1086 - acc: 0.9684 - val_loss: 0.0940 - val_acc

In [53]:
score = model3.evaluate(X_test, y_test, verbose=VERBOSE)
print("test score: ", score[0])
print("test accuracy: ", score[1])

10000/10000 [==============================] - 1s 54us/step
test score:  0.10480711509194261
test accuracy:  0.9777


### Try Adam optimizer

In [54]:
from keras.optimizers import Adam

In [55]:
OPTIMIZER = Adam()
model3.compile(loss = 'categorical_crossentropy', optimizer = OPTIMIZER, metrics = ['accuracy'])

In [56]:
model3.fit(X_train, y_train, batch_size=BATCH_SIZE, epochs=NP_EPOCH, verbose=VERBOSE, validation_split=VALIDATION_SPLIT)

Train on 48000 samples, validate on 12000 samples
Epoch 1/20
48000/48000 [==============================] - 3s 70us/step - loss: 0.0743 - acc: 0.9783 - val_loss: 0.0912 - val_acc: 0.9778
Epoch 2/20
48000/48000 [==============================] - 4s 84us/step - loss: 0.0702 - acc: 0.9782 - val_loss: 0.0901 - val_acc: 0.9779
Epoch 3/20
48000/48000 [==============================] - 3s 59us/step - loss: 0.0637 - acc: 0.9803 - val_loss: 0.0940 - val_acc: 0.9772
Epoch 4/20
48000/48000 [==============================] - 3s 64us/step - loss: 0.0632 - acc: 0.9807 - val_loss: 0.0917 - val_acc: 0.9770
Epoch 5/20
48000/48000 [==============================] - 3s 61us/step - loss: 0.0631 - acc: 0.9806 - val_loss: 0.0907 - val_acc: 0.9774
Epoch 6/20
48000/48000 [==============================] - 3s 66us/step - loss: 0.0595 - acc: 0.9818 - val_loss: 0.0889 - val_acc: 0.9770
Epoch 7/20
48000/48000 [==============================] - 3s 58us/step - loss: 0.0586 - acc: 0.9816 - val_loss: 0.0888 - val_acc

In [57]:
score = model3.evaluate(X_test, y_test, verbose=VERBOSE)
print("test score: ", score[0])
print("test accuracy: ", score[1])

10000/10000 [==============================] - 1s 54us/step
test score:  0.0800942157833626
test accuracy:  0.9806


Clearly, Adam optimizer gives the best test accuracy from all the algorithms